cmake/FindLibBpf.cmake

```diff
@@ -0,0 +1,59 @@
+# - Try to find libbpf
+# Once done this will define
+#
+#  LIBBPF_FOUND            - system has libbpf
+#  LIBBPF_INCLUDE_DIR      - the libbpf include directory
+#  LIBBPF_SOURCE_DIR       - the libbpf source directory
+#  LIBBPF_LIBRARIES        - link these to use libbpf
+
+#if (LIBBPF_LIBRARIES AND LIBBPF_INCLUDE_DIR AND LIBBPF_SOURCE_DIR)
+#  set (LibBpf_FIND_QUIETLY TRUE)
+#endif (LIBBPF_LIBRARIES AND LIBBPF_INCLUDE_DIR AND LIBBPF_SOURCE_DIR)
+
+find_path (LIBBPF_INCLUDE_DIR
+  NAMES
+    bpf/bpf.h
+    bpf/btf.h
+    bpf/libbpf.h
+
+  PATHS
+    /usr/include
+    /usr/local/include
+    /opt/local/include
+    /sw/include
+    ENV CPATH)
+
+file(GLOB libbpf_source_path /usr/src/debug/libbpf-*)
+
+find_path (LIBBPF_SOURCE_DIR
+  NAMES
+    src/bpf.c
+    src/bpf.h
+    src/libbpf.c
+    src/libbpf.h
+
+  PATHS
+    ${libbpf_source_path}
+    ENV CPATH
+)
```

Comment:
Which package will have bpf.c and libbpf.c?
From bcc perspective, we do not really need libbpf .c files,
the necessary header files for compilation and linking to libbpf.so should be good enough.

Reply:
when you set `CMAKE_USE_LIBBPF_PACKAGE`, you are building from libbpf rpm,
bpf.c and libbpf.c come from libbpf source package, check this hunk:

```
-file(GLOB libbpf_sources "libbpf/src/*.c")
+if(LIBBPF_FOUND)
+  # We need to include the libbpf packaged source to get the
+  # headers location, becaseu they are installed under 'bpf'
+  # directory, but the libbpf code references them localy
+
+  include_directories(${LIBBPF_INCLUDE_DIR}/bpf)
+
+  file(GLOB libbpf_sources "${LIBBPF_SOURCE_DIR}/src/*.c")
+  set(libbpf_uapi ${LIBBPF_SOURCE_DIR}/include/uapi/linux)
+else()
+  file(GLOB libbpf_sources "libbpf/src/*.c")
+  set(libbpf_uapi libbpf/include/uapi/linux})
+endif()

```

Reply:
it's libbpf-debugsource

Reply:
```
[yhs@localhost ~]$ sudo yum install libbpf-debugsource
Last metadata expiration check: 0:02:01 ago on Wed 30 Oct 2019 08:00:30 PM PDT.
No match for argument: libbpf-debugsource
Error: Unable to find a match: libbpf-debugsource
[yhs@localhost ~]$ uname -a
Linux localhost.localdomain 5.2.16-200.fc30.x86_64 #1 SMP Thu Sep 19 16:14:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
[yhs@localhost ~]$
```
Maybe I miss some setting?

Reply:
looks like you need to enable the updates repo, like:
`dnf install --enablerepo=updates-debuginfo libbpf-debugsource`

Reply:
Thanks. I can get the source now.

You are missing to add C++ API to bcc-shared.so library.

Also, the name bcc-shared.so does not sound great. It is hard to distinguish libbcc.so and libbcc-shared.so just by name. How about bcc-no-libbpf.so? a little bit long but captures what it is.

Could you have a test case when CMAKE_USE_LIBBPF_PACKAGE is used? The test case should exercise C++ API and should compile successfully on a host with libbpf-debugsource and other libbpf packages installed.

Reply:
Please also add some comments in the code to describe what packages are needed to
find all needed headers and sources. For example, people may not be aware of this one `dnf install --enablerepo=updates-debuginfo libbpf-debugsource`.

Reply:
> Thanks. I can get the source now.
> 
> You are missing to add C++ API to bcc-shared.so library.

ah ok, will check

> Also, the name bcc-shared.so does not sound great. It is hard to distinguish libbcc.so and libbcc-shared.so just by name. How about bcc-no-libbpf.so? a little bit long but captures what it is.

ok

> Could you have a test case when CMAKE_USE_LIBBPF_PACKAGE is used? The test case should exercise C++ API and should compile successfully on a host with libbpf-debugsource and other libbpf packages installed.

ok, thanks for review

Reply:
@yonghong-song I'm adding the test but I'm getting following error
when running test_libbcc:

```
 : CommandLine Error: Option 'no-x86-call-frame-opt' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```

just making sure it's an issue in my config  as google search suggest..

thanks

Reply:
Never seen this error before. This is related compilation options. Most likely it is a user error.

Reply:
@yonghong-song  I've managed to fix it by linking the libbcc.so dynamically,
which I think should be the way in any case..

I recently upgraded to new fedora that uses llvm-9, what llvm version are you compiling with?

Reply:
same thing happens for llvm 8 ... I pushed new code with fix for the test,
now both tests give same result